### PR TITLE
Refactor server game storage

### PR DESCRIPTION
**Pull request by khalidsharmarke**
Refactored game storage into class instances that allow for more modular and direct control over game state
edited socket validation to more accurately reflect which requests to drop

### CODE REVIEW

File: client/scripts/index.js [hunk 1]
```diff
@@ -20,6 +22,20 @@ function getPlayerColorFromNumber(player_number) {
 
 }
 
+function enableButtons(data){
```
Comment (willwen) [inline client/scripts/index.js]: why does this take the argument data ?

Comment (willwen) [reply client/scripts/index.js]: this also seems to modify the whiteboard
soo
maybe call the function
enableButtonsAndBoard

File: client/scripts/index.js [hunk 2]
```diff
@@ -1,3 +1,5 @@
+let pad = null
```
Comment (willwen) [inline client/scripts/index.js:3]: you can put a TODO here 
you made it global to ignore scoping
but eventually you can pass around the pad to each function that needs it to control the scope

Comment (willwen) [reply client/scripts/index.js:3]: and add a trello ticket to refactor `pad`'s scoping

File: client/scripts/index.js [hunk 3]
```diff
@@ -69,18 +86,19 @@ function game_init(pad) {
 // this function exports the whiteboard as JSON and sends it to the server
 function submit_stroke_to_server(player_number, pad) {
     const newBoardState = JSON.stringify(pad.toJSON())
-    console.log(newBoardState)
     socket.emit('gameplay_stroke', {
         player_number: player_number,
         new_board_state: encodeURIComponent(newBoardState)
     })
+    disableButtons()
```
Comment (willwen) [inline client/scripts/index.js]: can add a comment here, to explain in plain english
that after you hit submit, we disable all buttons for the user  until its their turn again

File: client/scripts/index.js [hunk 4]
```diff
@@ -69,18 +86,19 @@ function game_init(pad) {
...
 }
 
 // TO DO:
 // needs implementation
 function update_image_from_server(data) {
     console.log('new image received')
-    console.log(data)
+    let thing = decodeURIComponent(data.new_board_state)
```
Comment (willwen) [inline client/scripts/index.js]: horrible variable name

File: client/scripts/index.js [hunk 5]
```diff
@@ -92,6 +110,11 @@ requirejs(['responsive-sketchpad/sketchpad'],
     }
 )
 
+function displayRoomID(data){
+    console.log(data)
```
Comment (willwen) [inline client/scripts/index.js]: no need to log this anymore

Comment (willwen) [reply client/scripts/index.js]: why is it called data
the param should be called room_id

File: client/scripts/index.js [hunk 6]
```diff
@@ -108,4 +131,7 @@ socket.on('disconnect', reason => {
     window.location = '/'
 })
```
Comment (willwen) [inline client/scripts/index.js:127]: can add a comment here explaining this is where you do all the socket handlers:
ex:
```
// action handlers after receiving messages from server:
socket.on('room_id', displayRoomID)
socket.on('new_image', update_image_from_server)
socket.on('player_number', disableButtons)
socket.on('enable_turn', enableButtons)
```

File: index.js [hunk 7]
```diff
@@ -1,8 +1,10 @@
 const PORT = process.env.PORT || 8000
 
+const {set_of_active_games, generate_game_id, checkIfGameAlive, createNewGame, getGameFromSet} = require('./utils/game')
```
Comment (willwen) [inline index.js]: inconsistent variable naming
is it `_` or Snake case ?

File: index.js [hunk 8]
```diff
@@ -44,43 +45,20 @@ app.use('/', function (req, res, next) {
...
 })
 
 app.post('/join_game', urlencodedParser, (req, res)=>{
 	const game_id_to_join = Number(req.body.game_id)
-	console.log(list_of_active_games)
-	if(list_of_active_games.includes(game_id_to_join)){
-		// game exists
+
+    const gameAlive = checkIfGameAlive(game_id_to_join)
+    if (gameAlive){
```
Comment (willwen) [inline index.js:54]: the spacing and tab formatting  is really messed up here

File: index.js [hunk 9]
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
 // validates socket requests
 function validateSocket(socket_obj){
     const room_id = getRoomFromSocket(socket_obj)
-
+    const room = getGameFromSet(room_id)
```
Comment (willwen) [inline index.js]: technically dont want const here
you alter call room.addPlayer
which manipulates the room's player field

Comment (willwen) [reply index.js]: `getGameFromSet` is unclear
Set isnt clear enough
I would do:
`getGameObjFromRoomId`
or `getGameObj()`

File: index.js [hunk 10]
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
...
     // adds socket to room if client is missing room and if room exists
     // will drop connections on server restart
-    if (room_id !== null && !socket_obj.rooms.has(room_id) && list_of_active_games.includes(room_id)){
+    if (room_id !== null && !socket_obj.rooms.has(room_id) && room !== null){
         socket_obj.join(room_id)
+        room.addPlayer(socket_obj.id)
+        return {room_id, room}
```
Comment (willwen) [inline index.js]: single return point
so result = {room_id, room}


result = null


return result

File: index.js [hunk 11]
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
...
 }
 
+// there exists documentation for Socket IO Middleware
+// possible to implement before io.on connection
 io.on('connection', socket => {
-    const room_id = validateSocket(socket)
-    // TODO:
-    // here we should instantiate a game obj
-    // to track game state and stack history
+    const {room_id, room} = validateSocket(socket)
```
Comment (willwen) [inline index.js]: this function can return null, but you dont handle the null case ?

File: index.js [hunk 12]
```diff
@@ -108,30 +86,38 @@ function getRoomFromSocket(socket_obj){
...
+
+    socket.emit('room_id', room_id)
+    socket.emit('player_number', room.getPlayerNumber(socket.id))
+    // needed to enable turn for room creator
+    io.to(room.getCurrentPlayer()).emit('enable_turn', true)
 
     // update users in room for new image
     socket.on('gameplay_stroke', data => {
```
Comment (willwen) [inline index.js]: in this case data is the image data right ? 
so maybe jsut call it
`json_image_data`

File: utils/ActiveGames.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+class ActiveGames{
+    add(room_id, room_obj){
```
Comment (willwen) [inline utils/ActiveGames.js]: it should initialize an empty dictionary

```
class ActiveGames {
     activeGamesSet = {}


    add ( ) {
    ...
    }
}
```

File: utils/Room.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
+module.exports = class Room{
+    #private = 123
```
Comment (willwen) [inline utils/Room.js]: ?why is this here

File: utils/ActiveGames.js (hunk 13 above)
Comment (willwen) [reply utils/ActiveGames.js]: like in here
https://github.com/khalidsharmarke/A-Fake-Artist-Goes-To-New-York/pull/9/files#diff-6ce89e5b021636a347e8d3ed4e43055c9ba2357870bfa913c76c796dd6f68c40R4

File: utils/Room.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    constructor(){
+        this.listOfPlayers = []
+        this.nextPlayer = 0;
+    }
+    incrementPlayer(){
```
Comment (willwen) [inline utils/Room.js]: its incrementing the player number not the player
make the function name better

File: utils/Room.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
+module.exports = class Room{
+    #private = 123
+    constructor(){
+        this.listOfPlayers = []
+        this.nextPlayer = 0;
```
Comment (willwen) [inline utils/Room.js]: nextPlayerNumber

File: utils/Room.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    }
+    incrementPlayer(){
+        this.nextPlayer++;
+        if (this.listOfPlayers.length == this.nextPlayer){
+            this.nextPlayer = 0
+        }
+    }
+    nextTurn(){
```
Comment (willwen) [inline utils/Room.js:17]: next turn does not sound like it returns anything, so it should return null

you should make a separate function that gets Player

so you do:
```
game.nextTurn();
nextPlayerNumber = game.getNextPlayerNumber()
```

notice how the function names describes exactly what is going on

File: utils/Room.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+        const player = this.listOfPlayers[this.nextPlayer]
```
Comment (willwen) [inline utils/Room.js]: this is the player's socket, not the player

File: utils/Room.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    addPlayer(player_as_socket){
+        this.listOfPlayers.push(player_as_socket)
+    }
+    getPlayerNumber(player_as_socket){
+        return this.listOfPlayers.indexOf(player_as_socket)
+    }
+    getCurrentPlayer(){
+        return this.listOfPlayers[this.nextPlayer]
+    }
+    test(){
```
Comment (willwen) [inline utils/Room.js]: what

File: utils/Room.js [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+        }
+    }
+    nextTurn(){
+        const player = this.listOfPlayers[this.nextPlayer]
+        console.log(`player ${this.nextPlayer + 1} is drawing`)
+        this.incrementPlayer()
+        return player
+    }
+    addPlayer(player_as_socket){
+        this.listOfPlayers.push(player_as_socket)
```
Comment (willwen) [inline utils/Room.js]: why player_as_socket

instead of 
`playerSocketId`

File: utils/game.js [hunk 21]
```diff
@@ -0,0 +1,42 @@
+const Room = require('./Room')
+const ActiveGames = require('./ActiveGames')
+
+// TO DO:
+// implement remove game after no users present
+const set_of_active_games = new ActiveGames()
+
+
+function generate_game_id(){
```
Comment (willwen) [inline utils/game.js]: fix camel or snake case

**Review by willwen — COMMENTED** [review]
lots of naming issues

File: utils/ActiveGames.js (hunk 13 above)
Comment (khalidsharmarke) [reply utils/ActiveGames.js]: i disagree. this becomes redundant.
given the scope of the class as an implementation to store rooms in a singular obj
i think it best to avoid having to write ActiveGames.rooms['insert room id here']

File: utils/Room.js (hunk 14 above)
Comment (khalidsharmarke) [reply utils/Room.js]: i was experimenting with private class fields
its been removed

File: client/scripts/index.js [hunk 22]
```diff
@@ -53,59 +72,61 @@ function sketchpad_init(Sketchpad) {
...
+    disablePlayerActions()
 }
 
-// TO DO:
-// needs implementation
-function update_image_from_server(data) {
-    console.log('new image received')
-    console.log(data)
+// overwrites current pad with JSONimage from server
+function updateImageFromServer(data) {
```
Comment (willwen) [inline client/scripts/index.js]: data -> whiteboard_json_data

File: index.js [hunk 23]
```diff
@@ -28,59 +29,31 @@ const cookieOptions = {
...
-		}
-		else{
-			new_game_id = generate_game_id()
-		}
-		
-	}
-	console.log(list_of_active_games)
-
-    // Set cookie
+    // Set cookie`
```
Comment (willwen) [inline index.js:46]: typo

File: index.js [hunk 24]
```diff
@@ -89,11 +62,11 @@ app.post('/join_game', urlencodedParser, (req, res)=>{
...
-	}
+    }
 })
 
 // checks if client is assigned to room
-function getRoomFromSocket(socket_obj){
+function getRoomIDFromSocket(socket_obj){
     try{
         const parsed_cookie = cookie.parse(socket_obj.request.headers.cookie)
         const room_id = Number(parsed_cookie['room_id'])
```
Comment (willwen) [inline index.js:71]: single return statements

File: index.js [hunk 25]
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
 
 // validates socket requests
 function validateSocket(socket_obj){
-    const room_id = getRoomFromSocket(socket_obj)
-
+    let result = null
+    const room_id = getRoomIDFromSocket(socket_obj)
```
Comment (willwen) [inline index.js:83]: ?

File: index.js [hunk 26]
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
...
 
+// there exists documentation for Socket IO Middleware
+// possible to implement before io.on connection
 io.on('connection', socket => {
-    const room_id = validateSocket(socket)
-    // TODO:
-    // here we should instantiate a game obj
-    // to track game state and stack history
+    // validate all incoming connections
+    const {room_id, room} = validateSocket(socket)
```
Comment (willwen) [inline index.js]: getRoomFromSocket()

Comment (willwen) [reply index.js]: room id should be a field on room obj

Comment (willwen) [reply index.js]: move redundant 
`room.getCurrentPlayer()`
up at top

File: index.js [hunk 27]
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
...
+    // stops server-side operations on invalid socket connection
+    if (room_id == null || room == null) return
```
Comment (willwen) [inline index.js]: only need room==null check

File: index.js [hunk 28]
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
...
-        io.in(room_id).emit('new_image', data)
+    socket.on('gameplay_stroke', image_as_json => {
+        // send all clients in room new image
+        io.in(room_id).emit('new_image', image_as_json)
+        room.nextTurn()
+        io.to(room.getCurrentPlayer()).emit('enable_turn', true)
     });
 });
-   
+
```
Comment (willwen) [inline index.js]: check your newlines

File: utils/ActiveGames.js (hunk 13 above)
Comment (willwen) [reply utils/ActiveGames.js]: look up `extends object`

File: utils/Room.js [hunk 29]
```diff
@@ -0,0 +1,31 @@
+module.exports = class Room{
```
Comment (willwen) [inline utils/Room.js]: export at the end of the file

File: utils/Room.js [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+    constructor(){
+        this.listOfPlayers = []
+        this.nextPlayerNumber = 0
```
Comment (willwen) [inline utils/Room.js]: get current player number in

and next player number can be derived from current player

File: utils/game.js [hunk 31]
```diff
@@ -0,0 +1,39 @@
...
+const ActiveGames = require('./ActiveGames')
+
+const setOfActiveGames = new ActiveGames()
+
+function generateGameId(){
+	return Math.floor(Math.random() * 1000)
+}
+
+function checkIfGameExists(game_id){
+    return setOfActiveGames.has(game_id)
```
Comment (willwen) [inline utils/game.js]: game_id vs room_id? make it consistent

File: utils/game.js [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+function getRoomFromActiveSet(game_id){
+    if (setOfActiveGames.has(game_id)){
```
Comment (willwen) [inline utils/game.js]: checkIfGameExists

File: utils/game.js [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+
+function createNewGame(){
+    let result = undefined
+    // TODO:
+    // here we should instantiate a game obj
+    // to track game state and stack history
+    try {
+        const game_id = generateGameId()
+        let game = new Room(game_id)
+        setOfActiveGames.add(game_id, game)
```
Comment (willwen) [inline utils/game.js]: jsut add a game obj

File: client/scripts/index.js [hunk 34]
```diff
@@ -19,9 +23,23 @@ function getPlayerColorFromNumber(player_number) {
     return '#' + colors[player_number]
 
 }
+// allows player interations with interface
```
Comment (willwen) [inline client/scripts/index.js:26]: your PR still doesnt have a clear title
and doesnt have a body

File: client/scripts/index.js [hunk 35]
```diff
@@ -19,9 +23,23 @@ function getPlayerColorFromNumber(player_number) {
...
+function enablePlayerActions(data){
```
Comment (willwen) [inline client/scripts/index.js]: why is data in the args

File: client/scripts/index.js [hunk 36]
```diff
@@ -53,59 +72,61 @@ function sketchpad_init(Sketchpad) {
...
-        submit_stroke_to_server(player_number, pad)
+        submitStrokeToServer(player_number, pad)
     };
-
 }
 
 // after a player has submitted their stroke,
 // this function exports the whiteboard as JSON and sends it to the server
-function submit_stroke_to_server(player_number, pad) {
+function submitStrokeToServer(player_number, pad) {
```
Comment (willwen) [inline client/scripts/index.js]: this function makes it sound like you are passing a stroke to the server
but the arg is pad
so either change the function to be
submit pad to server
OR
modify the param to be the strokeJson data instead of pad

File: index.js (hunk 23 above)
Comment (willwen) [reply index.js:46]: why is this resolved if the typo is still there

File: index.js [hunk 37]
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
 
 // validates socket requests
 function validateSocket(socket_obj){
-    const room_id = getRoomFromSocket(socket_obj)
-
+    let result = null
+    const room_id = getRoomIDFromSocket(socket_obj)
+    const room = getRoomFromActiveSet(room_id)
```
Comment (willwen) [inline index.js]: `getRoomObjFromID`

File: index.js [hunk 38]
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
 
 // validates socket requests
 function validateSocket(socket_obj){
-    const room_id = getRoomFromSocket(socket_obj)
-
+    let result = null
```
Comment (willwen) [inline index.js:82]: this function doesnt validate socket
it returns a room

File: index.js [hunk 39]
```diff
@@ -107,31 +80,43 @@ function getRoomFromSocket(socket_obj){
...
+    const room_id = getRoomIDFromSocket(socket_obj)
+    const room = getRoomFromActiveSet(room_id)
     // adds socket to room if client is missing room and if room exists
     // will drop connections on server restart
-    if (room_id !== null && !socket_obj.rooms.has(room_id) && list_of_active_games.includes(room_id)){
+    if (room_id !== null && !socket_obj.rooms.has(room_id) && room !== null){
         socket_obj.join(room_id)
+        room.addPlayer(socket_obj.id)
+        result = {room_id, room}
```
Comment (willwen) [inline index.js]: why return both room_id and room

File: utils/Room.js [hunk 40]
```diff
@@ -0,0 +1,31 @@
...
+    }
+    addPlayer(player_as_socket_id){
+        this.listOfPlayers.push(player_as_socket_id)
+        return null
+    }
+    getPlayerNumber(player_as_socket_id){
+        return this.listOfPlayers.indexOf(player_as_socket_id)
+    }
+    getCurrentPlayer(){
+        return this.listOfPlayers[this.nextPlayerNumber]
```
Comment (willwen) [inline utils/Room.js]: i dont like the variable names here
isee you are leveraging the fact that arrays are 0 indexed,
but it should really look like:
```
getCurrentPlayer(){
     return this.listOfCurerntPlayers[this.currentPlayerNumber + 1 ] // +1 because array index starts at 0
}
```

File: client/scripts/index.js (hunk 35 above)
Comment (khalidsharmarke) [reply client/scripts/index.js]: i think it important to denote that the function is able to receive data from the server.
Im still thinking theres an implementation where the server passes the pad object back to the client
with that being how we control scope

**Review by willwen — APPROVED** [review]
